### PR TITLE
WASM: Jit loop body yield restore 2.0

### PR DESCRIPTION
**Pull request by Cellule**
OS#13885461
This integrates the reverted changes from #3838
However, instead of trying to guess if the register is a yield register, explicitly save all the yield register to detect them in the irbuilder.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/chakracore/4240)

### CODE REVIEW

**Comment by Cellule** [comment]
ping

File: lib/Backend/IRBuilderAsmJs.cpp [hunk 1]
```diff
@@ -445,7 +432,7 @@ IRBuilderAsmJs::AddStatementBoundary(uint statementIndex, uint offset)
     return m_statementReader.MoveNextStatementBoundary();
 }
 
-Js::RegSlot IRBuilderAsmJs::GetTypedRegFromRegSlot(Js::RegSlot reg, WAsmJs::Types type)
+uint32 IRBuilderAsmJs::GetTypedRegFromRegSlot(Js::RegSlot reg, WAsmJs::Types type)
```
Comment (MikeHolman) [inline lib/Backend/IRBuilderAsmJs.cpp]: nit: why this change?

File: lib/Backend/IRBuilderAsmJs.cpp [hunk 2]
```diff
@@ -599,19 +586,84 @@ IRBuilderAsmJs::RegIsVar(Js::RegSlot reg)
...
+                    GetJitLoopBodyData().SetRegIsYield(reg);
+                    EnsureLoopBodyAsmJsStoreSlot(reg, type);
+                }
+            }
+        }
+        else
+#endif
+        {
+            // In Asm.js there is no Yields outside of the loop, if we're here
+            // It means we are probably returning a value from the loop
```
Comment (MikeHolman) [inline lib/Backend/IRBuilderAsmJs.cpp]: only probably?

File: lib/WasmReader/WasmByteCodeGenerator.cpp [hunk 3]
```diff
@@ -938,10 +938,24 @@ EmitInfo WasmBytecodeGenerator::EmitLoop()
     Js::ByteCodeLabel loopHeadLabel = m_writer->DefineLabel();
     Js::ByteCodeLabel loopLandingPadLabel = m_writer->DefineLabel();
 
-    uint32 loopId = m_writer->EnterLoop(loopHeadLabel);
+    // Push possibly yielding loop label before capturing all the yielding registers
+    BlockInfo implicitBlockInfo = PushLabel(loopTailLabel);
+
+    // Save all the possible blocks the loop could yield to
```
Comment (MikeHolman) [inline lib/WasmReader/WasmByteCodeGenerator.cpp]: maybe change this comment? You are saving all the registers that are internal to the loop, there's nothing really about blocks

File: lib/Backend/IRBuilderAsmJs.cpp (hunk 1 above)
Comment (Cellule) [reply lib/Backend/IRBuilderAsmJs.cpp]: I wanted to differentiate an IR Reg from a ByteCode typed Reg. 
I thought it didn't really make sense to treat those as RegSlot

File: lib/Backend/IRBuilderAsmJs.cpp (hunk 2 above)
Comment (Cellule) [reply lib/Backend/IRBuilderAsmJs.cpp]: Humm you're right, it should be definitively a return, I'll fix the comment

File: lib/WasmReader/WasmByteCodeGenerator.cpp (hunk 3 above)
Comment (Cellule) [reply lib/WasmReader/WasmByteCodeGenerator.cpp]: Right, I forgot to update the comment in my last iteration thanks
